src/common/logging.cpp

```diff
@@ -28,64 +28,270 @@
 
 #include "logging.hpp"
 
+#include <stdio.h>
+#include <stdlib.h>
+#include <stdarg.h>
 #include <assert.h>
 #include <errno.h>
 #include <stdarg.h>
 #include <stdint.h>
 #include <string.h>
 #include <syslog.h>
+#include <sys/time.h>
 
 static int        sLevel = LOG_INFO;
 static const char HEX_CHARS[] = "0123456789abcdef";
 
+static unsigned msecs_start;
+static bool     log_col0 = true; /* we start at col0 */
+static FILE    *log_fp;
+static bool     syslog_enabled = true;
+static bool     syslog_opened = false;
+
+#define LOGFLAG_syslog 1
+#define LOGFLAG_file   2
+
+/** Set/Clear syslog enable flag */
+void otbrLogEnableSyslog(bool b)
+{
+    syslog_enabled = b;
+}
+
+/** Enable logging to a specific file */
+void otbrLogSetFilename(const char *filename)
+{
+    log_fp = fopen(filename, "w");
+    if (log_fp == NULL)
+    {
+        fprintf(stderr, "Cannot open log file: %s\n", filename);
+        perror(filename);
+        exit(EXIT_FAILURE);
+    }
+}
+
+/** Get the current debug log level */
+int  otbrLogGetLevel(void)
+{
+    return sLevel;
+}
+
+/** Set the debug log level */
+void otbrLogSetLevel(int aLevel)
+{
+    assert(aLevel >= LOG_EMERG && aLevel <= LOG_DEBUG);
+    sLevel = aLevel;
+}
+
+/** Determine if we should not or not log, and if so where to */
+static int do_log(int aLevel)
+{
+    int r;
+
+    assert(aLevel >= LOG_EMERG && aLevel <= LOG_DEBUG);
+
+    r = 0;
+
+    if (syslog_opened && syslog_enabled && (aLevel <= sLevel))
+    {
+        r = r | LOGFLAG_syslog;
+    }
+
+    /* if someone has turned on the seperate file the most likely
+     * situation is they are debugging a problem, or need a extra
+     * information. In this case, we do not test the log level.
+     */
+    if (log_fp != NULL)
+    {
+        r = r | LOGFLAG_file;
+    }
+    return r;
+}
+
+
+/** return the time, in milliseconds since application start */
+static unsigned msecs_now(void)
+{
+    struct timeval tv;
+    uint64_t       v;
+
+    gettimeofday(&tv, NULL);
+
+    v = tv.tv_sec;
+    v = v * 1000;
+    v = v + (tv.tv_usec / 1000);
+
+    if (msecs_start == 0)
```

Comment:
With the current implementation, the returned time actually depends on the first call to `msecs_now()`.  That is, if `msecs_now()` is firstly called `10s` after the program runs, it will return `0` instead of expected `10 * 1000`. I think this is different from what the comment indicates: `/** return the time, in milliseconds since application start */`. On the other hand, `otbrLogInit()` is supposed to be called at the beginning of the application. Initializing `msecs_start ` inside `otbrLogInit()` would provide more stable output.